platform/app/src/routes/Mode/Mode.tsx

```diff
@@ -77,31 +105,47 @@ function defaultRouteInit(
     });
   });
 
-  // The hanging protocol matching service is fairly expensive to run multiple
-  // times, and doesn't allow partial matches to be made (it will simply fail
-  // to display anything if a required match fails), so we wait here until all metadata
-  // is retrieved (which will synchronously trigger the display set creation)
-  // until we run the hanging protocol matching service.
-
-  Promise.allSettled(allRetrieves).then(() => {
+  Promise.allSettled(allRetrieves).then(promises => {
     log.timeEnd(TimingEnum.STUDY_TO_DISPLAY_SETS);
     log.time(TimingEnum.DISPLAY_SETS_TO_FIRST_IMAGE);
     log.time(TimingEnum.DISPLAY_SETS_TO_ALL_IMAGES);
-    const displaySets = displaySetService.getActiveDisplaySets();
 
-    if (!displaySets || !displaySets.length) {
-      return;
+    const allPromises = [];
+    const remainingPromises = [];
+
+    function startRemainingPromises(remainingPromises) {
+      remainingPromises.forEach(p => p.forEach(p => p.start()));
     }
 
-    // Gets the studies list to use
-    const studies = getStudies(studyInstanceUIDs, displaySets);
+    function runHP() {
+      const { unsubscribe } = hangingProtocolService.subscribe(
+        hangingProtocolService.EVENTS.PROTOCOL_CHANGED,
+        () => {
+          startRemainingPromises(remainingPromises);
+          unsubscribe();
+        }
+      );
+      applyHangingProtocol();
+    }
 
-    // study being displayed, and is thus the "active" study.
-    const activeStudy = studies[0];
+    promises.forEach(promise => {
+      const retrieveSeriesMetadataPromise = promise.value;
+      if (Array.isArray(retrieveSeriesMetadataPromise)) {
+        const { requiredSeries, remaining } = hangingProtocolService.filterSeriesRequiredForRun(
+          hangingProtocolId,
+          retrieveSeriesMetadataPromise
+        );
+        const requiredSeriesPromises = requiredSeries.map(promise => promise.start());
+        allPromises.push(Promise.allSettled(requiredSeriesPromises));
+        remainingPromises.push(remaining);
+      }
+    });
 
-    // run the hanging protocol matching on the displaySets with the predefined
-    // hanging protocol in the mode configuration
-    hangingProtocolService.run({ studies, activeStudy, displaySets }, hangingProtocolId);
+    if (allPromises.length) {
+      Promise.allSettled(allPromises).then(() => runHP());
```

Comment:
Could you just start the remaining promises here?  That way they should be done by the time the HP applies.

Reply:
Updated, thanks.